File: leet_code_75/max_profit.js
const maxProfit = (prices) => {
    let minPrice = Infinity;
    let profit = 0;

    for (let i = 0; i < prices.length; i++) {
        let price = prices[i]
        if (minPrice > price) {
            minPrice = price
        } else if (minPrice < price) {
            const newProfit = price - minPrice;

            if (newProfit > profit) profit = newProfit;
        }
    }

    return profit;
};

// Input: [7, 1, 5, 3, 6, 4]
// Output: 5
// Explanation: Buy on day 2(price = 1) and sell on day 5(price = 6), profit = 6 - 1 = 5.
// Not 7 - 1 = 6, as selling price needs to be larger than buying price.

// Input: [7, 6, 4, 3, 1]
// Output: 0
// Explanation: In this case, no transaction is done, i.e.max profit = 0.

File: leet_code/python/arrays/max_profit.py
# O(n) time, O(1) space
def maxProfit(self, prices):
        if len(prices) <= 1:
                return 0

        max_profit = float("-inf")
        min_price = float("inf")

        for price in prices:
            min_price = min(min_price, price)
            current_profit = price - min_price
            max_profit = max(max_profit, current_profit)

        return max_profit
